l2geth/rollup/sync_service.go

```diff
@@ -1215,12 +1233,82 @@ func (s *SyncService) syncTransactionRange(start, end uint64, backend Backend) e
 	return nil
 }
 
+// syncTransactionFromQueue will sync the earliest transaction from an external message queue
+func (s *SyncService) syncTransactionFromQueue() error {
+	// we don't drop messages unless they're already applied
+	cb := func(ctx context.Context, msg QueueSubscriberMessage) {
+		var (
+			txMeta types.TransactionMeta
+			tx     types.Transaction
+		)
+		if err := json.Unmarshal(msg.Data(), &txMeta); err != nil {
+			log.Error("Failed to unmarshal logged TransactionMeta", "msg", err)
+			msg.Nack()
+			return
+		}
+		if err := rlp.DecodeBytes(txMeta.RawTransaction, &tx); err != nil {
+			log.Error("decoding raw transaction failed", "msg", err)
+			msg.Nack()
+			return
+		}
+		if txMeta.L1BlockNumber == nil || txMeta.L1Timestamp == 0 {
+			log.Error("Missing required queued transaction fields", "msg", string(msg.Data()))
+			msg.Nack()
+			return
+		}
+
+		if readTx, _, _, _ := rawdb.ReadTransaction(s.db, tx.Hash()); readTx != nil {
+			msg.Ack()
+			return
+		}
+
+		if err := s.applyTransactionToTip(&tx); err != nil {
```

Comment:
Do we have an ordering guarantee? And what about a delivery guarantee?

Reply:
Yeah. The ordering is enforced by PubSub via external configuration. And the transactions published by a Sequencer are ordered - since a SendTx fails if we're unable to log it in PubSub.

Delivery guarantees are also handled by PubSub via external configuration. On infra, published messages will be available for at least several days. Which is enough time for a failover. Messages may be delivered more than once, so we check the db to see if we've already applied the transaction. If so, we acknowledge the message (advance the subscription queue offset) and move on.

Reply:
We can assert that infra is setup correctly in code and the PubSub subscription has message ordering enabled. But this requires extra GCP permissions than needed to subscribe to messages.